nav2_safety_nodes/src/nav2_safety_node.cpp

```diff
@@ -0,0 +1,250 @@
+    /// string of polygon points and returns a polygon vector
+    nav2_util::makeVectorPointsFromString(safety_polygon_, safety_zone);
+  } else {
+    // Polygon provided but invalid, so stay with the radius
+    RCLCPP_ERROR(
+      logger_, "The safety_polygon is invalid: \"%s\" :) ",
+      safety_polygon_.c_str());
+  }
+}
+
+void
+SafetyZone::initTransforms()
+{
+  RCLCPP_INFO(get_logger(), "initTransforms");
+  // Initialize transform listener and broadcaster
+  tf2_ = std::make_unique<tf2_ros::Buffer>(get_clock());
+  auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+    get_node_base_interface(), get_node_timers_interface());
+  tf2_->setCreateTimerInterface(timer_interface);
+  tf2_listener_ = std::make_unique<tf2_ros::TransformListener>(*tf2_);
+}
+
+// Publishers and subscribers
+void
+SafetyZone::initPubSub()
+{
+  RCLCPP_INFO(logger_, "initPubSub");
+  // Create the publishers and subscribers
+  safety_polygon_pub_ = create_publisher<geometry_msgs::msg::PolygonStamped>(
+    "published_polygon", rclcpp::SystemDefaultsQoS());
+  // Pointcloud publisher
+  point_cloud_pub_ = create_publisher<sensor_msgs::msg::PointCloud2>(
+    "cloud", rclcpp::SensorDataQoS());
+  // Velocity publisher
+  speed_limit_pub_ = create_publisher<nav2_msgs::msg::SpeedLimit>(
+    speed_limit_topic_, rclcpp::QoS(10));
+
+  // Multiple Laserscan subscribers
+  scan_subscribers_.reserve(scan_topics_.size());
+  RCLCPP_INFO(logger_, "Subscribing to scan topics");
+  for (unsigned int i = 0; i < scan_topics_.size(); i++) {
+    scan_subscribers_.push_back(
+      create_subscription<sensor_msgs::msg::LaserScan>(
+        scan_topics_[i].c_str(), rclcpp::SystemDefaultsQoS(),
+        std::bind(&SafetyZone::laserCallback, this, std::placeholders::_1)));
+  }
+  // Timer -> 10hzs
+  timer_ = create_wall_timer(
+    100ms, std::bind(&SafetyZone::timerCallback, this));
+  RCLCPP_INFO(logger_, "Subscribed to scan topics");
+}
+
+void
+SafetyZone::laserCallback(
+  const sensor_msgs::msg::LaserScan::SharedPtr message)
+{
+  // project the laser into a point cloud
+  const sensor_msgs::msg::PointCloud2::SharedPtr cloud;
+
+  try {
+    projector_.projectLaser(*message, *cloud);
+  } catch (...) {
+    RCLCPP_WARN(
+      logger_,
+      "Unable to project laser into cloud");
+  }
+
+  // Transform the cloud if necessary
+  if (!base_frame_.empty() && cloud->header.frame_id != base_frame_) {
+    try {
+      *cloud = tf2_->transform(*cloud, base_frame_, tf2::durationFromSec(tf_tolerance_));
+      sensor_data.push(*cloud);
+    } catch (tf2::TransformException & ex) {
+      RCLCPP_ERROR_STREAM(logger_, "Transform failure: " << ex.what());
+      return;
+    }
+  } else if (cloud->header.frame_id == base_frame_) {
+    sensor_data.push(*cloud);
+  }
+}
+
+double
+SafetyZone::dotProduct(
+  const Eigen::Vector3d & pt1,
+  const Eigen::Vector3d & pt2)
+{
+  return pt1[0] * pt2[1] - pt1[1] * pt2[0];
```

Comment:
Use eigens `dot()` function for vector 3d. Remove this function

Reply:
I don't know the reason but the thing is when I used the inbuilt `dot()` function, I was getting different results for dot product, which in turn resulted in false detection of points, so I continued using the function created.

Reply:
That is because you use the Vector3d type, try the Vector2d type instead. The result is different because you populated the `z` components of the vector which are being used in the computation.

Reply:
No, it isn't working either with Vector2d, giving different results

Reply:
why?

Reply:
This is the comparison between results for 5 test points, dot product of each point with all 5 edges of polygon using both
| dotProduct  | dot() |
| ------------- | ------------- |
| 0.75 0.5 0 0 0.25  | 0.25 0.25 1 0 -0.25  |
| -0.1 0.1 0.9 0.4 0.2  | 0.2 -0.2 0.2 0.45 0.6  |
| 0.25 0 0 0.5 0.75  | 0.75 0.25 0 0 0.25  |
| 0 0.25 1 0.25 0   | -0 -0.25 0.5 0.5 0.5  |
| 0.35 0.3 0.4 0.2 0.25   | 0.25 0.05 0.6 0.2 0.15  |

Reply:
I'm not asking about how its different, I'm asking why its different, please assess it.

Reply:
It's because in our dot product function we aren't actually calculating dot product it just substitutes the point in the equation of the line segment for determining which side of the line segment does the point lie.